experimental/Schemes/Schemes.jl

```diff
@@ -0,0 +1,2587 @@
+    (k,a) = divides_power( proj_loc(phi_prime(f_i)), proj_loc(pullback_from_root(U)(g)) )
+    phi_of_f *= lift(a)*t^k
+  end
+  push!( images, phi_of_f )
+  phi = AlgebraHomomorphism( R_t, S_t, images )
+  h.pullback = phi
+  return phi
+end
+
+# Construct the fractional representation of the morphism 
+# from the explicit algebra homomorphism. 
+@Markdown.doc """
+    function imgs_frac(f::AffSchMorphism)
+
+Returns the images xᵢ of the generators of the ambient ring R of the root Y 
+of the codomain V of f : U → V in their fractional representation. 
+"""
+function imgs_frac(f::AffSchMorphism)
+  # first check if this variable is already cached
+  if isdefined(f, :imgs_frac)
+    return f.imgs_frac
+  end
+  # if both forms of the morphisms are not to be found, it is not defined at all.
+  if !isdefined( f, :pullback )
+    error( "Neither the fractional representation, nor the pullback is defined for this morphism" )
+  end
+
+  # start reconstructing the fraction representation from 
+  # the explicit ring homomorphism
+  phi = pullback(f)
+  # Set up the codomain of a lift of the pullback phi to the ambient ring of 
+  # the domain of f. This is a localization of a *free* polynomial ring, 
+  # considered as a subalgebra of the fraction field. 
+  P = ambient_ring(root(domain(f)))
+  F = FractionField(P)
+  den = div_denoms(domain(f))
+  d = length(den)
+  n = length(gens(P))
+  # prepare a list of the images of the generators 
+  fracs = [ F(x) for x in gens(P) ]
+  push!( fracs, prod( [ 1//g for g in den ] ))
+  R = ambient_ring(root(codomain(f)))
+  imgs_frac = elem_type(F)[]
+  h = compose( pullback_from_root(codomain(f)), phi )
+  for g in gens(R)
+    push!(imgs_frac,evaluate(h(g), fracs))
+  end
+  f.imgs_frac = imgs_frac
+  return imgs_frac
+end
+
+@Markdown.doc """
+    function identity_morphism( X::Spec ) -> AffSchMorphism
+
+Returns the identity morphism of a Spec.
+"""
+function identity_morphism( X::Spec )
+  R = ambient_ring(X)
+  phi = AlgebraHomomorphism( R, R, gens(R) )
+  f = AffSchMorphism(X,X,phi)
+  return f 
+end
+
+inclusion_in_root( X::Spec ) = identity_morphism( X )
+inclusion_in_parent( X::Spec ) = identity_morphism( X )
+
+@Markdown.doc """
+    function identity_morphism( D::SpecPrincipalOpen ) -> AffSchMorphism 
+
+Returns the identity morphism of a SpecPrincipalOpen.
+"""
+function identity_morphism( D::SpecPrincipalOpen )
+  R = ambient_ring(D)
+  phi = AlgebraHomomorphism( R, R, gens(R) )
+  f = AffSchMorphism(D,D,phi)
+  return f
+end
+
+#####################################################################
+# Composition of affine scheme morphisms. 
+# Consists basically of composition of the associated pullback 
+# homomorphisms.
+#
+@Markdown.doc """
+    function compose( f::AffSchMorphism, g::AffSchMorphism ) 
+
+Returns the composition f ∘ g of two morphisms of affine schemes. 
+NOTE: The ordering of the arguments follows the mathematical convention, 
+unlike other functions for composition in Oscar! 
```

Comment:
Err, what? That's (a) completely unclear (in how far does OSCAR's `compose` not follow a mathematical convention, and which is that???), and secondly sounds like a big no-no (namely, as if you are deliberately violating the specification of `compose`, which would be really bad, no matter that some code already does this, see issue #690)